### ISSUE TITLE
Remote Server Instantly closes

### ISSUE THREAD

**Issue by zolem** [joonty/vdebug#209]
I have the problem that whenever I try to debug in windows from gvim and use a VBOX server on a bridged network it doesn't work.

If I try to specify a server ip address no matter what it is I will instantly get the message "Connection to the debugger has been closed"

My servers xdebug settings are:
```INI
zend_extension = /usr/lib/php5/20121212/xdebug.so
xdebug.remote_enable=on
xdebug.remote_handler=dbgp
xdebug.remote_port=9000
xdebug.remote_connect_back=1
```
Everything for vdebug settings is default. If I leave it that way it waits for a connection and times out obviously because my server is not on localhost. If I specify any address at all to the server option it immediately fails.

**Comment by zolem** [comment]
I am still having this problem sadly but I don't think anyone looks at these issues anymore.

It appears to me that if you set a server as anything but localhost vdebug does something that breaks. This is strange behavior because every other debugger I have used for PHP doesn't care about the remote address. All it does is listen on port 9000 for xdebug to initiate a connection. Once a connection is made commands are sent like normal. Seems like simple behavior to me but I could be wrong.

I just don't understand why I would need to set a server IP for the remote connection.

**Comment by zolem** [comment]
Well I decided to comb through the code and see where the server is actually being specified and the fix appears to be pretty easy. Hopefully this can be implemented in a future release. I will try a pull request and see if it works out.

Here is what I found in the meantime: https://github.com/joonty/vdebug/blob/master/plugin/python/vdebug/dbgp.py#L415
```py
serv.bind((self.host, self.port))
```
The above code makes the socket listen specifically to the IP address specified in the server option variable. This is completely unnecessary from what I can tell. It would make more sense to just accept a connection from anywhere which is what I believe netbeans does.
Here is the fix:
```py
serv.bind(('', self.port))
```

**Comment by joonty** [comment]
Hi @zolem,

Firstly, very sorry for the long wait you've had! I still look at the issues from time to time, and there's a big pile of stuff in my todo list (but sadly a lack of time - hopefully this will change as I'm [rounding up a few people to maintain the plugin](http://blog.joncairns.com/2015/01/looking-for-another-vdebug-maintainer/)).

The issue you're having is due to the fact that the server option in vdebug isn't actually the IP or host for the machine that's running the PHP script, but the IP address that Vdebug will bind to. This is important because by default, localhost (the default value for the server option) is 127.0.0.1. Therefore vdebug starts a socket server listening on 127.0.0.1, and the PHP script with xdebug acts as the client. However, 127.0.0.1 will only allow local connections. If you want a PHP script to connect from another machine, you would probably have to bind the Vdebug server to something like 0.0.0.0, which is all interfaces. Then, probably after setting the correct firewall settings on the router, you would be able to tunnel from the client to the vdebug server and connect to port 9000.

Bit of a rambling explanation - I'm sure that could have been done better. But basically it is necessary! The confusion is in what acts as the server and what acts as the client. Vdebug is the server and your PHP script is the client.

Does that help? Let me know if you're still having problems getting your setup working.

Thanks

**Comment by zolem** [comment]
Hi @joonty,
This does make sense however in the python script if you want to open the server to allow connection from any IP you simply don't specify an IP to bind it to. Therefore if any machine (xdebug client) tries to connect to vdebug via port 9000 then the debugger works just fine over local or remote debugging as long as the machine can reach your computer running vdebug. This simple change I suggested above allows debugging from both local and remote computers without the need for setting any options associated with where xdebug is comming from.

The problem I had after doing this change on my local vdebug machine was that pathmaps didn't work at all. Although I am not sure if pathmaps were functioning for remote debugging beforehand anyway. In the refresh method where the file being debugged should have been called it would call a method
```py
self.cur_file = vdebug.util.RemoteFilePath(stack[0].get('filename'))
```
Then it would call:
```py
selef.ui.set_source_position(\
    self.cur_file,\
    self.cur_lineno)
```
The class vdebug.util.RemoteFilePath actually did nothing which I thought was strange. See below:
```py
class RemoteFilePath(FilePath):
    def _create_remote(self, f)
        """Create the file name valid for the remote server.

        Uses the "local_path" and "remote_path" options.
        """
        return f
```
Although I don't 100% understand python because it is not a language I work in this appears to be an empty method. How the above method call actually calls the _create_remote method I am not sure but it appears to do so in the inherited __init__ method. And by adding the code from the FilePath class under the _create_remote method it seems to work. Of course in order to pull the proper local file you would need to change it to call LocalFilePath instead of RemoteFilePath and then add the code from FilePath class under _create_local method to the LocalFilePathClass.

I understand that RemoteFilePath and LocalFilePath classes inherit from FilePath, but for some reason calling the super method did not work for me. There is probably a better implementation but someone with more python experience should certainly create that.

In short the changes that I made in my pull request works great for me. I am able to debug just fine my remote php installations without trouble. Whether or not you wish to implement these changes is up to you but in the meantime I will simply use my fork of the project until they are implemented. Thanks for your help.

**Comment by joonty** [comment]
You convinced me ;)

Added this to the `dev` branch - soon to be released in 1.5.0. It binds to all addresses by default, and I've updated the documentation to reflect that.

Thanks for sticking with me on this one.

**Comment by zolem** [comment]
Hey this is great! I also made a few random cosmetic changes on my fork you should check out. You might like them too.
